evm/p2p/peer.py

```diff
@@ -350,36 +364,96 @@ def __init__(self, remote, privkey, reader, writer, aes_secret, mac_secret,
+        messages from the same remote peer.
         """
-        with (yield from self._header_fetching_lock):
-            self.logger.info("fetch_headers({}, reorg_depth={}) called".format(
-                block_number, reorg_depth))
-            head_number = self.chaindb.get_canonical_head().block_number
-            head_number -= reorg_depth
-            announced_head_number = block_number
-            while announced_head_number > head_number:
-                request_id = gen_request_id()
-                target_head = min(head_number + MAX_HEADERS_FETCH, announced_head_number)
-                self.les_proto.send_get_block_headers(
-                    target_head, MAX_HEADERS_FETCH, request_id, reverse=True)
-                reply = yield from self.wait_for_reply(request_id)
-                for header in reversed(reply['headers']):
-                    self.chaindb.persist_header_to_db(header)
-                    # FIXME: The reference to the canonical chain head should not be stored in the
-                    # chaindb as that is shared by all peers -- instead, that should be kept on a
-                    # peer-specific db.
-                    self.chaindb.add_block_number_to_hash_lookup(header)
-                    self.chaindb.set_canonical_head(header.hash)
-                    head_number = header.block_number
-                self.logger.info("synced headers up to block {}".format(head_number))
+        if self.last_synced is None:
+            # It's the first time we hear from this peer, need to figure out which headers to
+            # get from it.  We can't simply fetch headers starting from our current head
+            # number because there may have been a chain reorg, so we fetch some headers prior
+            # to our head from the peer, and insert any missing ones in our DB, essentially
+            # making our canonical chain identical to the peer's up to
+            # chain_head.block_number.
+            chain_head = self.chaindb.get_canonical_head()
+            self.logger.info(
+                "First announcement by {}; head_number=={}, local chain_head=={}".format(
+                    self, head_number, chain_head.block_number))
+            oldest_ancestor_to_consider = max(
+                0, chain_head.block_number - self.max_headers_fetch + 1)
+            headers = yield from self._fetch_headers_starting_at(oldest_ancestor_to_consider)
+            if len(headers) == 0 or not self.chaindb.header_exists(headers[0].hash):
+                self.logger.warn("No common ancestors between us and remote, disconnecting")
+                self.disconnect(DisconnectReason.other)
+                return
+            for header in headers:
+                self.chaindb.persist_header_to_db(header)
+            self.last_synced = chain_head.block_number
+        else:
+            self.last_synced = self.last_synced - reorg_depth
+
+        while self.last_synced < head_number:
```

Comment:
At this point do we *know* that this peer has a header of at least `head_number` height?  I glanced but couldn't find it.  If not, do we need to add that validation to avoid this loop never terminating?

Reply:
Yeah, `head_number` is the head just announced by the remote, so it should be able to provide us all headers up to (inclusive) that